---
title: "Estimate marker effects for fatty acids"
author: "Malachy Campbell"
date: "12/16/2020"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, root.dir = "/Users/malachycampbell/Documents/Dropbox/Work/Oats/DivPanel_Met_HH/statTargetQC/flashr")
```

## Boxplot of FAs in DP and Elite panels.
```{r, echo = T, eval = F}
rm(list = ls())

DPdata <- readRDS("../../../TargetedMet/DP/FinalData.rds")$rawData
Eldata <- readRDS("../../../TargetedMet/Elite/FinalData.rds")$RawData

DPdata <- DPdata[c(7, 21:29)]
Eldata <- Eldata[c(6, 2, 12:20)]

DPdata <- reshape2::melt(DPdata, id.vars = "plot")
Eldata <- reshape2::melt(Eldata, id.vars = c("plot", "env"))
Eldata$order <- NA
Eldata$order[Eldata$variable == "C14.0"] <- 1
Eldata$order[Eldata$variable == "C16.0"] <- 2
Eldata$order[Eldata$variable == "C16.1"] <- 3
Eldata$order[Eldata$variable == "C18.0"] <- 4
Eldata$order[Eldata$variable == "C18.1"] <- 5
Eldata$order[Eldata$variable == "C18.2"] <- 6
Eldata$order[Eldata$variable == "C18.3"] <- 7
Eldata$order[Eldata$variable == "C20.0"] <- 8
Eldata$order[Eldata$variable == "C20.1"] <- 9
Eldata <- Eldata[order(Eldata$order) ,]

pdf("../../../KernelPred_Front/Manuscript/SuppFigs/FA_boxplots.pdf", h = 6, w = 5.2)
nf = layout(rbind(c(1,1), c(2,2)))
par(mar = c(3.5, 4.5, 1, 0.5))
boxplot(log(value) ~ variable, data = DPdata, ylab = expression("log(mg"~"g"^-1*")" ), 
        las = 2, cex = 0.5, medlwd = 1.25, cex.axis = 0.7, xlab = "", 
        names = sub("\\.", ":", unique(DPdata$variable)))

mtext("A", 2, adj=8, las=1, padj=-12, font=1, cex=0.8)

boxplot(log(value) ~ env*order, data = Eldata, ylab = expression("log(mg"~"g"^-1*")" ), 
        las = 2, cex = 0.5, medlwd = 1.25, xlab = "", xaxt = "n", 
        at = c(1,2,3, 5,6,7, 9,10,11, 
               13,14,15, 17,18,19, 21,22,23, 
               25,26,27, 29,30,31, 33,34,35),
        col = wesanderson::wes_palette("Chevalier1")[2:4])

axis(1, at = c(2, 6, 10, 14, 18, 22, 26, 30, 34), 
     labels = sub("\\.", ":", unique(Eldata$variable)), las = 2, cex.axis = 0.7, font = 1)

legend("topright", legend = c("MN", "SD", "WI"), 
       fill = wesanderson::wes_palette("Chevalier1")[2:4], bty = "n", cex = 0.65)

mtext("B", 2, adj=8, las=1, padj=-12, font=1, cex=0.8)
dev.off()
```

## Prep FA data (targeted)

```{r, echo = T, eval = F}
rm(list = ls())

# load markers
mrks <- data.table::fread("../../GenoData/results/allMrks_imp_SNPs_genomat_73257x539.txt", sep = "\t", header = F)
mrkNames <- read.table("../../GenoData/SNPnames_allMrks_imp_SNPs_genomat_73257x539.txt", header = F)
mapData <- read.csv("../../../TrevorOats/genotypes/data/T_AHOY_OC3_Monkey.csv")

mapData <- data.frame(marker = mapData$Locus, chr = mapData$Chr, pos = mapData$Dist)
mapDataAnch <- mapData[mapData$marker %in% mrkNames$V1 ,]
mapData <- rbind(mapDataAnch,
                 data.frame(marker = mrkNames[! mrkNames$V1 %in% mapData$marker ,], 
                            chr = "Unk", 
                            pos = 1:length(mrkNames[! mrkNames$V1 %in% mapData$marker ,])))

############################
# Cleaning phenotypic data #
############################
# Load phenotypes for FAs
data <- readRDS("../../../TargetedMet/DP/FinalData.rds")
BLUPs <- data$BLUPs$drBLUPs

lineNames <- unique(data$rawData[c("Line", "GID")])
lineNames <- lineNames[match(row.names(BLUPs), lineNames$Line) ,]
sum(lineNames$Line == row.names(BLUPs))
row.names(BLUPs) <- lineNames$GID

#file used to convert T3 IDs to GID
IDconv <- read.csv("../../../TrevorOats/genotypes/results/diversity_t3_names.csv")
BLUPs <- BLUPs[!is.na(BLUPs[,1]) ,]; dim(BLUPs)
IDconv <- IDconv[IDconv$GID %in% row.names(BLUPs) ,]; dim(IDconv) #368 x 3

IDconv <- IDconv[IDconv$t3_name %in% mrks$V1 ,]; dim(IDconv) #336 x 3
mrks <- mrks[mrks$V1 %in% IDconv$t3_name ,]; dim(mrks) #336 x 73258
BLUPs <- BLUPs[row.names(BLUPs) %in% IDconv$GID ,]; dim(BLUPs) #336 9

#reorder IDconv to match BLUPs, and reorder and rename marker data to match BLUPs
IDconv <- IDconv[match(row.names(BLUPs), IDconv$GID) ,]; sum(row.names(BLUPs) == IDconv$GID) #336
mrks <- mrks[match(IDconv$t3_name, mrks$V1) ,]; sum(IDconv$t3_name == mrks$V1) #336
mrks$V1 <- IDconv$GID

#Replace elements in Res with new data
data$BLUPs$drBLUPs_mrk <- BLUPs

mrks$V1 <- NULL #drop names from the marker data
mrks <- as.matrix(mrks)
mrks <- mrks + 1

###############################
# Filtering at a marker-level #
###############################
#Filter based on MAF
freq <- colMeans(mrks)/2
maf <- ifelse(freq > 0.5, 1-freq, freq)
maf.index <- which(maf < 0.05); length(maf.index) #11208 markers failed MAF threshold
mrks <- mrks[,-maf.index]; dim(mrks) #335 62002
mrkNames <- mrkNames[-maf.index, drop = F ,]

#Load map data. For unanchored markers create a data frame, where chromosome is Unk and postion is an index
mapData <- read.table("../../../TrevorOats/genotypes/data/T_AHOY_OC3_Monkey.csv", header = T, sep = ",")
mapData <- data.frame(marker = mapData$Locus, chr = mapData$Chr, pos = mapData$Dist)
mapDataAnch <- mapData[mapData$marker %in% mrkNames$V1 ,]
mapData <- rbind(mapDataAnch,
                 data.frame(marker = mrkNames[! mrkNames$V1 %in% mapData$marker ,], 
                            chr = "Unk", 
                            pos = 1:length(mrkNames[! mrkNames$V1 %in% mapData$marker ,])))
mapData <- mapData[match(mrkNames$V1, mapData$marker) ,]

saveRDS(list(phenotypes = BLUPs,
             mrks = mrks,
             map = mapData), "../../../KernelPred_Front/InputData_FAs.rds")

BLUPs <- merge(BLUPs, IDconv[c("GID", "t3_name")], by.x = 0, by.y = "GID")
BLUPs <- data.frame(line = BLUPs$t3_name, BLUPs[2:10])
write.csv(BLUPs, "../../../KernelPred_Front/Manuscript/SuppFiles/DP_FA.csv", row.names = F)
```

## Estimate h2 for each compound.
```{r, echo = T, eval = F}
library(corrplot)

rm(list = ls())

inputData <- readRDS("../../../KernelPred_Front/InputData_FAs.rds")

phenos <- inputData$phenotypes

Zsc <- scale(x = inputData$mrks, center = T, scale = T)
GRM <- tcrossprod(Zsc)/ncol(Zsc)
GRM <- GRM + diag(nrow(Zsc))*0.001

Ys <- scale(phenos, center = T, scale = T)

ETA <- list(list(K = GRM, model = "RKHS"))
h2 <- array()
for(i in 1:ncol(Ys)){
  fm <- BGLR::BGLR(y = Ys[,i], ETA = ETA, nIter = 20000, burnIn = 5000, thin = 5)
  h2[i] <- fm$ETA[[1]]$varU/(fm$varE + fm$ETA[[1]]$varU)
}

names(h2) <- colnames(phenos)
print(h2)
#     C14.0     C16.0     C16.1     C18.0     C18.1     C18.2     C18.3     C20.0     C20.1 
# 0.5894352 0.6927498 0.6385693 0.7756895 0.7471989 0.6819376 0.3350327 0.7327735 0.7040376 


# fit using a multitrait model
rm(list = ls())

inputData <- readRDS("../../../KernelPred_Front/InputData_FAs.rds")

phenos <- inputData$phenotypes

Zsc <- scale(x = inputData$mrks, center = T, scale = T)
GRM <- tcrossprod(Zsc)/ncol(Zsc)
GRM <- GRM + diag(nrow(Zsc))*0.001

Ys <- scale(phenos, center = T, scale = T)

ETA <- list(list(K = GRM, model = "RKHS", ))
ETA <- list(Ga = list(K= GRM, model = "RKHS", Cov=list(df0 = ncol(Ys))))
mtMod <- BGLR::Multitrait(y = Ys, ETA = ETA, nIter = 20000, burnIn = 5000)

saveRDS(mtMod, "../../../KernelPred_Front/MTM_cov.rds")

mtMod <- readRDS("../../../KernelPred_Front/MTM_cov.rds")
gCor <- cov2cor(mtMod$ETA[[1]]$Cov$Omega)
corMat <- cov2cor(mtMod$ETA[[1]]$Cov$Omega + mtMod$resCov$R) 
corMat[upper.tri(corMat, diag = F)] <- gCor[upper.tri(gCor)]
diag(corMat) <- diag(mtMod$ETA[[1]]$Cov$Omega)/diag(mtMod$ETA[[1]]$Cov$Omega + mtMod$resCov$R)
colnames(corMat) <- row.names(corMat) <- sub("\\.", ":", colnames(corMat))

COLS <- colorRampPalette(c(wesanderson::wes_palette("Zissou1")[1],
                           "#FFFFFF", 
                           wesanderson::wes_palette("Zissou1")[5]))
Pmat <- corMat
Pmat[upper.tri(Pmat, diag = F)] <- NA
Pmat[lower.tri(Pmat, diag = F)] <- NA

cex.before <- par("cex")
pdf("../../../KernelPred_Front/Manuscript/Figs/CorMat.pdf", h = 4.2, w = 4.2)
par(cex = 0.7)
corrplot::corrplot(corr = corMat, method = "circle", is.corr = T,
                   order = "original", tl.col = "black", tl.cex = par("cex"), 
                   cl.cex = par("cex"), number.cex = par("cex"), col = COLS(1000), 
                   p.mat = Pmat, sig.level = -1, insig = "p-value", mar = c(0, 0.1, 0.1, 0.1))
dev.off()

par(cex = cex.before)
```


## Estimate $\beta$ via BayesianRR, Bayesian LASSO, Bayes A, Bayes B, Bayes Cpi
```{r, echo = T, eval = F}
library(BGLR)

rm(list = ls())

InputData <- readRDS("../InputData_FAs.rds") #for mapping information

mrks <- InputData$mrks
BLUPs <- InputData$phenotypes

Zsc <- scale(x = mrks, center = T, scale = T)
GRM <- tcrossprod(Zsc)/ncol(Zsc)
GRM <- GRM + diag(nrow(Zsc))*0.001

rownames(mrks) <- paste0("GID_", row.names(BLUPs))
colnames(GRM) <- row.names(GRM) <- paste0("GID_", row.names(BLUPs))
mrkEff <- matrix(NA, ncol = ncol(BLUPs), nrow = ncol(mrks))
modDets <- NULL

for (i in 1:ncol(BLUPs)){
  Y <- scale(BLUPs[,i], center = T, scale = F)
  
  #BRR
  if(!dir.exists("/work/walia/malachyc/O_m/front/FA/BRR")) dir.create("/work/walia/malachyc/O_m/front/FA/BRR")
  ETA <- list(list(X = mrks, model = "BRR"))
  BRRmodel <- BGLR(y = Y, ETA = ETA, 
                 nIter = 20000, burnIn = 5000, 
                 thin = 5, 
                 saveAt = paste0("/work/walia/malachyc/O_m/front/FA/BRR/", annotations[i]))
  saveRDS(BRRmodel, paste0("/work/walia/malachyc/O_m/front/FA/BRR/BRR_", annotations[i], ".rds"))
  
  #BL
  if(!dir.exists("/work/walia/malachyc/O_m/front/FA/BL")) dir.create("/work/walia/malachyc/O_m/front/FA/BL")
  ETA <- list(list(X = mrks, model = "BL"))
  BLmodel <- BGLR(y = Y, ETA = ETA, 
                 nIter = 20000, burnIn = 5000, 
                 thin = 5, 
                 saveAt = paste0("/work/walia/malachyc/O_m/front/FA/BL/", annotations[i]))
  saveRDS(BLmodel, paste0("/work/walia/malachyc/O_m/front/FA/BL/BL_", annotations[i], ".rds"))
  
  #Bayes A
  if(!dir.exists("/work/walia/malachyc/O_m/front/FA/BayesA")) dir.create("/work/walia/malachyc/O_m/front/FA/BayesA")
  ETA <- list(list(X = mrks, model = "BayesA"))
  BayesAmodel <- BGLR(y = Y, ETA = ETA, 
                 nIter = 20000, burnIn = 5000, 
                 thin = 5, 
                 saveAt = paste0("/work/walia/malachyc/O_m/front/FA/BayesA/", annotationsn[i]))
  saveRDS(BayesAmodel, paste0("/work/walia/malachyc/O_m/front/FA/BayesA/BayesA_", annotations[i], ".rds"))
  
  #Bayes B
  if(!dir.exists("/work/walia/malachyc/O_m/front/FA/BayesB")) dir.create("/work/walia/malachyc/O_m/front/FA/BayesB")
  ETA <- list(list(X = mrks, model = "BayesB"))
  BayesBmodel <- BGLR(y = Y, ETA = ETA, 
                 nIter = 20000, burnIn = 5000, 
                 thin = 5, 
                 saveAt = paste0("/work/walia/malachyc/O_m/front/FA/BayesB/", annotations[i]))
  saveRDS(BayesBmodel, paste0("/work/walia/malachyc/O_m/front/FA/BayesB/BayesB_", annotations[i], ".rds"))
  
  #Bayes C
  if(!dir.exists("/work/walia/malachyc/O_m/front/FA/BayesC")) dir.create("/work/walia/malachyc/O_m/front/FA/BayesC")
  ETA <- list(list(X = mrks, model = "BayesC"))
  BayesCmodel <- BGLR(y = Y, ETA = ETA, 
                 nIter = 20000, burnIn = 5000, 
                 thin = 5, 
                 saveAt = paste0("/work/walia/malachyc/O_m/front/FA/BayesC/", annotations[i]))
  saveRDS(BayesCmodel, paste0("/work/walia/malachyc/O_m/front/FA/BayesC/BayesC_", annotations[i], ".rds"))
}
```


## Prep data for Elite population.
```{r, echo = T, eval = F}
rm(list = ls())


# Load phenotypes for FAs
data <- readRDS("../../../TargetedMet/Elite/FinalData.rds")
BLUPs <- rbind(data.frame(line = data$BLUPs$MN$drBLUPs$line, env = "MN", data$BLUPs$MN$drBLUPs[2:ncol(data$BLUPs$MN$drBLUPs)]),
               data.frame(line = data$BLUPs$SD$drBLUPs$line, env = "SD", data$BLUPs$SD$drBLUPs[2:ncol(data$BLUPs$MN$drBLUPs)]),
               data.frame(line = data$BLUPs$WI$drBLUPs$line, env = "WI", data$BLUPs$WI$drBLUPs[2:ncol(data$BLUPs$MN$drBLUPs)]))
colNames <- colnames(BLUPs)

###################################################
# Pad missing lines for each environment with NAs #
###################################################
BLUPs <- reshape2::melt(BLUPs, id.vars = c("line", "env"))
BLUPs <- reshape2::dcast(BLUPs, line ~ env + variable, id.vars = "value")
# Remove checks (no BLUPs because they were treated as fixed)
chks <- apply(BLUPs[,2:ncol(BLUPs)], 1, function(x) sum(is.na(x))); names(chks) <- BLUPs$line
print(names(chks[chks == 27]))
# [1] "Corral__ck" "Deon__ck"   "Ogle__ck"
BLUPs <- BLUPs[-which(BLUPs$line %in% names(chks[chks == 27])) ,]
BLUPs$line <- droplevels(BLUPs$line)
BLUPs <- reshape2::melt(BLUPs, id.vars = "line")
metaBLUPs <- do.call(rbind, strsplit(as.character(BLUPs$variable), "_", 2))
dim(BLUPs)[1] == dim(metaBLUPs)[1]
BLUPs <- data.frame(line = BLUPs$line, metaBLUPs, BLUPs$value)
colnames(BLUPs)[2:3] <- c("env", "cmpd") 
BLUPs <- reshape2::dcast(BLUPs, line + env ~ cmpd, value.var = "BLUPs.value")
BLUPs <- BLUPs[, match(colNames, colnames(BLUPs))]

######

mrks <- data.table::fread("../../GenoData/results/allMrks_imp_SNPs_genomat_73257x539.txt", sep = "\t", header = F)
mrkNames <- read.table("../../GenoData/SNPnames_allMrks_imp_SNPs_genomat_73257x539.txt", header = F)
mapData <- read.csv("../../../TrevorOats/genotypes/data/T_AHOY_OC3_Monkey.csv")

mapData <- data.frame(marker = mapData$Locus, chr = mapData$Chr, pos = mapData$Dist)
mapDataAnch <- mapData[mapData$marker %in% mrkNames$V1 ,]
mapData <- rbind(mapDataAnch,
                 data.frame(marker = mrkNames[! mrkNames$V1 %in% mapData$marker ,], 
                            chr = "Unk", 
                            pos = 1:length(mrkNames[! mrkNames$V1 %in% mapData$marker ,])))

eliteLines <- read.csv("../../../TrevorOats/genotypes/results/elite_t3_names.csv")
eliteLines <- eliteLines[eliteLines$t3_name %in% mrks$V1 ,]; dim(eliteLines) #218 x 2
mrks <- mrks[mrks$V1 %in% eliteLines$t3_name ,]
eliteLines <- eliteLines[match(mrks$V1, eliteLines$t3_name) ,]
sum(mrks$V1 == eliteLines$t3_name) == nrow(mrks)
mrks$V1 <- eliteLines$fieldbook_name
#mrks$V1 <- NULL

dim(BLUPs) #676  11
length(unique(BLUPs$line)) #232
BLUPs <- BLUPs[BLUPs$line %in% mrks$V1 ,]; BLUPs$line <- droplevels(BLUPs$line); dim(BLUPs)
BLUPs$line <- as.character(BLUPs$line)
BLUPs <- BLUPs[order(BLUPs$env, BLUPs$line) ,]
length(unique(BLUPs$line)) #213
summary(data.frame(table(BLUPs$line)))
mrks <- mrks[mrks$V1 %in% BLUPs$line ,]
mrks$V1 <- droplevels(mrks$V1)
dim(mrks) #213 73257

mrks <- mrks[order(mrks$V1) ,]
mrkMat <- mrks[order(mrks$V1) ,][, 2:ncol(mrks)]
mrkMat <- as.matrix(mrkMat)
mrkMat <- mrkMat + 1

###############################
# Filtering at a marker-level #
###############################
#Filter based on MAF
freq <- colMeans(mrkMat)/2
maf <- ifelse(freq > 0.5, 1-freq, freq)
maf.index <- which(maf < 0.05); length(maf.index) #15135 markers failed MAF threshold
mrkMat <- mrkMat[,-maf.index]; dim(mrkMat) #213 58122
mrkNames <- mrkNames[-maf.index, drop = F ,]

#Load map data. For unanchored markers create a data frame, where chromosome is Unk and postion is an index
mapData <- read.table("../../../TrevorOats/genotypes/data/T_AHOY_OC3_Monkey.csv", header = T, sep = ",")
mapData <- data.frame(marker = mapData$Locus, chr = mapData$Chr, pos = mapData$Dist)
mapDataAnch <- mapData[mapData$marker %in% mrkNames$V1 ,]
mapData <- rbind(mapDataAnch,
                 data.frame(marker = mrkNames[! mrkNames$V1 %in% mapData$marker ,], 
                            chr = "Unk", 
                            pos = 1:length(mrkNames[! mrkNames$V1 %in% mapData$marker ,])))
mapData <- mapData[match(mrkNames$V1, mapData$marker) ,]

colnames(mrkMat) <- mrkNames$V1
mrks$V1 <- as.character(mrks$V1)
mrks <- data.frame(V1 = mrks$V1, mrkMat)
sum(unique(BLUPs$line) == mrks$V1)

saveRDS(list(phenotypes = BLUPs,
             mrks = mrks,
             mapData = mapData), "../../../KernelPred_Front/InputData_FAs_elite.rds")

```

## Create matrix of marker effects
```{r, echo = T, eval = F}
rm(list = ls())

InputDP <- readRDS("../../../KernelPred_Front/InputData_FAs.rds")
InputElite <- readRDS("../../../KernelPred_Front/InputData_FAs_elite.rds")

eff16.0 <- matrix(NA, ncol = 5, nrow = nrow(InputDP$map))
eff18.1 <- matrix(NA, ncol = 5, nrow = nrow(InputDP$map))
eff18.2 <- matrix(NA, ncol = 5, nrow = nrow(InputDP$map))
eff18.0 <- matrix(NA, ncol = 5, nrow = nrow(InputDP$map))
eff18.3 <- matrix(NA, ncol = 5, nrow = nrow(InputDP$map))
eff14.0 <- matrix(NA, ncol = 5, nrow = nrow(InputDP$map))
eff20.0 <- matrix(NA, ncol = 5, nrow = nrow(InputDP$map))
eff20.1 <- matrix(NA, ncol = 5, nrow = nrow(InputDP$map))
eff16.1 <- matrix(NA, ncol = 5, nrow = nrow(InputDP$map))

BayesAlph <- c("BRR", "BL", "BayesA", "BayesB", "BayesC")
for (i in 1:length(BayesAlph)){
  eff16.0[,i] <- readRDS(paste0("../../../KernelPred_Front/FA/", BayesAlph[i] , "/", BayesAlph[i], "_C16.0.rds"))$ETA[[1]]$b
  eff18.1[,i] <- readRDS(paste0("../../../KernelPred_Front/FA/", BayesAlph[i] , "/", BayesAlph[i], "_C18.1.rds"))$ETA[[1]]$b
  eff18.2[,i] <- readRDS(paste0("../../../KernelPred_Front/FA/", BayesAlph[i] , "/", BayesAlph[i], "_C18.2.rds"))$ETA[[1]]$b
  eff18.0[,i] <- readRDS(paste0("../../../KernelPred_Front/FA/", BayesAlph[i] , "/", BayesAlph[i], "_C18.0.rds"))$ETA[[1]]$b
  eff18.3[,i] <- readRDS(paste0("../../../KernelPred_Front/FA/", BayesAlph[i] , "/", BayesAlph[i], "_C18.3.rds"))$ETA[[1]]$b
  eff14.0[,i] <- readRDS(paste0("../../../KernelPred_Front/FA/", BayesAlph[i] , "/", BayesAlph[i], "_C14.0.rds"))$ETA[[1]]$b
  eff20.0[,i] <- readRDS(paste0("../../../KernelPred_Front/FA/", BayesAlph[i] , "/", BayesAlph[i], "_C20.0.rds"))$ETA[[1]]$b
  eff20.1[,i] <- readRDS(paste0("../../../KernelPred_Front/FA/", BayesAlph[i] , "/", BayesAlph[i], "_C20.1.rds"))$ETA[[1]]$b
  eff16.1[,i] <- readRDS(paste0("../../../KernelPred_Front/FA/", BayesAlph[i] , "/", BayesAlph[i], "_C16.1.rds"))$ETA[[1]]$b
  }

row.names(eff16.0) <- row.names(eff18.1) <- row.names(eff18.2) <- row.names(eff18.0) <- row.names(eff18.3) <- row.names(eff14.0) <- row.names(eff20.0) <- row.names(eff20.1) <- row.names(eff16.1) <- InputDP$map$marker

InputElite$mrkEff <- list(C16.0 = eff16.0,
                          C18.1 = eff18.1,
                          C18.2 = eff18.2,
                          C18.0 = eff18.0,
                          C18.3 = eff18.3,
                          C14.0 = eff14.0,
                          C20.0 = eff20.0,
                          C20.1 = eff20.1,
                          C16.1 = eff16.1)

# Indexes for folds
phenos <- InputElite$phenotypes
Lines <- unique(phenos$line)

flds <- matrix(NA, ncol = 50, nrow = length(Lines))
fldIndx <- rep(c(1:5), each = ceiling(length(Lines)/5))
for(i in 1:50){
  set.seed((451758 + i))
  flds[,i] <- fldIndx[sample(1:length(fldIndx))][1:length(Lines)]
}
row.names(flds) <- Lines
InputElite$fldIndx <- flds
saveRDS(InputElite, "../../../KernelPred_Front/InputData_FAs_elite.rds")
```

Get the amount of genetic variance explained by top markers for 18.1
```{r, echo = T, eval = F}
rm(list = ls())

InputDP <- readRDS("../../../KernelPred_Front/InputData_FAs.rds")
C18.1 <- readRDS("../../../KernelPred_Front/FA/BayesB/BayesB_C18.1.rds")
mrks <- InputDP$mrks
colnames(mrks) <- InputDP$map$marker

maf <- colMeans(mrks)/2
betas <- C18.1$ETA[[1]]$b
varBeta <- C18.1$ETA[[1]]$varB
names(varBeta) <- names(betas) <- names(maf)

genVar <- sum(maf * (1-maf) * varBeta)
eVar <- C18.1$varE
totalVar <- (genVar + eVar)

betas <- betas[order(abs(betas), decreasing = T)]
genVar_1 <- maf[names(maf) %in% names(betas)[1]] * (1 - maf[names(maf) %in% names(betas)[1]]) * varBeta[names(varBeta) %in% names(betas)[1]]

genVar_1 / totalVar * 100 # 0.05660077
genVar_1 / genVar * 100 # 0.08795561
```
